services/imageproxy/atmos_camo.go

```diff
@@ -45,22 +54,32 @@ func (backend *AtmosCamoBackend) GetImageDirect(imageURL string) (io.ReadCloser,
 
 func (backend *AtmosCamoBackend) getAtmosCamoImageURL(imageURL string) string {
 	cfg := *backend.proxy.ConfigService.Config()
-	siteURL := *cfg.ServiceSettings.SiteURL
-	proxyURL := *cfg.ImageProxySettings.RemoteImageProxyURL
 	options := *cfg.ImageProxySettings.RemoteImageProxyOptions
 
-	return getAtmosCamoImageURL(imageURL, siteURL, proxyURL, options)
-}
+	if imageURL == "" || backend.siteURL == nil {
+		return imageURL
+	}
+
+	// Parse url, bypass if error in parsing.
+	parsedURL, err := url.Parse(imageURL)
+	if err != nil {
+		return imageURL
+	}
 
-func getAtmosCamoImageURL(imageURL, siteURL, proxyURL, options string) string {
-	// Don't proxy blank images, relative URLs, absolute URLs on this server, or URLs that are already going through the proxy
-	if imageURL == "" || imageURL[0] == '/' || (siteURL != "" && strings.HasPrefix(imageURL, siteURL)) || strings.HasPrefix(imageURL, proxyURL) {
+	// If host is same as siteURL host/ remoteURL host, or it's a relative URL, return.
+	if parsedURL.Host == backend.siteURL.Host || parsedURL.Host == "" || parsedURL.Host == backend.remoteURL.Host {
 		return imageURL
```

Comment:
In fact, we should never return `imageURL`, only `parsedURL.String()` -- and even then we should always fill in the blanks first: if `parsedURL.Host` is empty we should set it to `backend.siteURL.Host`. Same for `Scheme`. That's because browsers will parse `https:///example.com` as pointing to the host `example.com`, whereas go sees an empty host name.

To quote Filippo on this: "any security property that relies on parser alignment is doomed" and "you don't parse and check, you rebuild and compare".

Reply:
>  if parsedURL.Host is empty we should set it to backend.siteURL.Host

There is a test against this which verifies that relative URLs should bypass the proxy. Are we sure we want to break this behavior? I am 0/5 on any of it, just confirming.

Reply:
Ah wait, you just want the host to be filled with the siteURL host right? It will still bypass, but the URL will be complete.

Reply:
That's it, yeah.